# EDA

In [137]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.feature_selection import f_classif
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


In [138]:
df = pd.read_csv("https://raw.githubusercontent.com/4GeeksAcademy/regularized-linear-regression-project-tutorial/main/demographic_health_data.csv")
df.head()


fips  TOT_POP    0-9  0-9 y/o % of total pop  19-Oct  \
0  1001    55601   6787               12.206615    7637   
1  1003   218022  24757               11.355276   26913   
2  1005    24881   2732               10.980266    2960   
3  1007    22400   2456               10.964286    2596   
4  1009    57840   7095               12.266598    7570   

   10-19 y/o % of total pop  20-29  20-29 y/o % of total pop  30-39  \
0                 13.735364   6878                 12.370281   7089   
1                 12.344167  23579                 10.814964  25213   
2                 11.896628   3268                 13.134520   3201   
3                 11.589286   3029                 13.522321   3113   
4                 13.087828   6742                 11.656293   6884   

   30-39 y/o % of total pop  ...  COPD_number  diabetes_prevalence  \
0                 12.749771  ...         3644                 12.9   
1                 11.564429  ...        14692                 12.0   
2                 12.865239  ...         2373                 19.7   
3                 13.897321  ...         1789                 14.1   
4                 11.901798  ...         4661                 13.5   

   diabetes_Lower 95% CI  diabetes_Upper 95% CI  diabetes_number  \
0                   11.9                   13.8             5462   
1                   11.0                   13.1            20520   
2                   18.6                   20.6             3870   
3                   13.2                   14.9             2511   
4                   12.6                   14.5             6017   

   CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  CKD_number  \
0             3.1               2.9               3.3        1326   
1             3.2               3.0               3.5        5479   
2             4.5               4.2               4.8         887   
3             3.3               3.1               3.6         595   
4             3.4               3.2               3.7        1507   

   Urban_rural_code  
0                 3  
1                 4  
2                 6  
3                 2  
4                 2  

[5 rows x 108 columns]

In [139]:
df.shape

(3140, 108)

The dataset is made up of 108 columns, of which we have to reduce in order to carry out the study, as analysing all of them would be an extensive and unproductive task.

In [140]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3140 entries, 0 to 3139
Columns: 108 entries, fips to Urban_rural_code
dtypes: float64(61), int64(45), object(2)
memory usage: 2.6+ MB


In [141]:
# Configures Pandas to display full DataFrames without truncation
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

In [142]:
# View all columns without truncation
df.head()

fips  TOT_POP    0-9  0-9 y/o % of total pop  19-Oct  \
0  1001    55601   6787               12.206615    7637   
1  1003   218022  24757               11.355276   26913   
2  1005    24881   2732               10.980266    2960   
3  1007    22400   2456               10.964286    2596   
4  1009    57840   7095               12.266598    7570   

   10-19 y/o % of total pop  20-29  20-29 y/o % of total pop  30-39  \
0                 13.735364   6878                 12.370281   7089   
1                 12.344167  23579                 10.814964  25213   
2                 11.896628   3268                 13.134520   3201   
3                 11.589286   3029                 13.522321   3113   
4                 13.087828   6742                 11.656293   6884   

   30-39 y/o % of total pop  40-49  40-49 y/o % of total pop  50-59  \
0                 12.749771   7582                 13.636445   7738   
1                 11.564429  27338                 12.539102  29986   
2                 12.865239   3074                 12.354809   3278   
3                 13.897321   3038                 13.562500   3115   
4                 11.901798   7474                 12.921853   7844   

   50-59 y/o % of total pop  60-69  60-69 y/o % of total pop  70-79  \
0                 13.917016   5826                 10.478229   4050   
1                 13.753658  29932                 13.728890  20936   
2                 13.174712   3076                 12.362847   2244   
3                 13.906250   2545                 11.361607   1723   
4                 13.561549   6965                 12.041840   4931   

   70-79 y/o % of total pop   80+  80+ y/o % of total pop  White-alone pop  \
0                  7.284042  2014                3.622237            42660   
1                  9.602701  9368                4.296814           190301   
2                  9.018930  1048                4.212049            12209   
3                  7.691964   785                3.504464            17211   
4                  8.525242  2335                4.036999            55456   

   % White-alone  Black-alone pop  % Black-alone  \
0      76.725239            10915      19.630942   
1      87.285228            19492       8.940382   
2      49.069571            12042      48.398376   
3      76.834821             4770      21.294643   
4      95.878285              950       1.642462   

   Native American/American Indian-alone pop  % NA/AI-alone  Asian-alone pop  \
0                                        267       0.480207              681   
1                                       1684       0.772399             2508   
2                                        164       0.659137              113   
3                                         98       0.437500               53   
4                                        378       0.653527              185   

   % Asian-alone  Hawaiian/Pacific Islander-alone pop  % Hawaiian/PI-alone  \
0       1.224798                                   62             0.111509   
1       1.150343                                  146             0.066966   
2       0.454162                                   46             0.184880   
3       0.236607                                   26             0.116071   
4       0.319848                                   70             0.121024   

   Two or more races pop  % Two or more races  POP_ESTIMATE_2018  \
0                   1016             1.827305              55601   
1                   3891             1.784682             218022   
2                    307             1.233873              24881   
3                    242             1.080357              22400   
4                    801             1.384855              57840   

   N_POP_CHG_2018  GQ_ESTIMATES_2018  R_birth_2018  R_death_2018  \
0             158                455          11.8           9.6   
1            5403               2190          10.5          10.3   
2            -277     

In [143]:
# Resets the value of pandas to display truncated dataframes
pd.reset_option('all')

/tmp/ipykernel_5509/4261215448.py:2: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  pd.reset_option('all')


In [144]:
# Obtain null values
df.isnull().sum()

fips                      0
TOT_POP                   0
0-9                       0
0-9 y/o % of total pop    0
19-Oct                    0
                         ..
CKD_prevalence            0
CKD_Lower 95% CI          0
CKD_Upper 95% CI          0
CKD_number                0
Urban_rural_code          0
Length: 108, dtype: int64

In [145]:
# Obtain duplicates
df.duplicated().sum()

0

In [146]:
# Get the columns with data type "object"
object_columns = df.select_dtypes(include=['object']).columns
object_columns


Index(['COUNTY_NAME', 'STATE_NAME'], dtype='object')

In [147]:
# Converted the categoricals to numericals
for col in object_columns:
    df[col] = pd.factorize(df[col])[0]

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3140 entries, 0 to 3139
Columns: 108 entries, fips to Urban_rural_code
dtypes: float64(61), int64(47)
memory usage: 2.6 MB


In [148]:
# Get data types of columns
data_types = df.dtypes

# Initialize StandardScaler
scaler = StandardScaler()

# Scale all variables, including those that were factorized
norm_features = scaler.fit_transform(df)

# Create a new DataFrame with all variables scaled
total_data_scal = pd.DataFrame(norm_features, index=df.index, columns=df.columns)
total_data_scal["diabetes_prevalence"] = df["diabetes_prevalence"]

# Show the first rows of the new DataFrame
total_data_scal.head()



fips   TOT_POP       0-9  0-9 y/o % of total pop    19-Oct  \
0 -1.940874 -0.145679 -0.142421                0.158006 -0.135556   
1 -1.940742  0.341296  0.287476               -0.242861  0.320383   
2 -1.940610 -0.237785 -0.239429               -0.419441 -0.246181   
3 -1.940478 -0.245223 -0.246032               -0.426966 -0.254791   
4 -1.940346 -0.138966 -0.135053                0.186249 -0.137140   

   10-19 y/o % of total pop     20-29  20-29 y/o % of total pop     30-39  \
0                  0.573496 -0.153144                  0.027610 -0.139384   
1                 -0.193107  0.183774                 -0.469965  0.230620   
2                 -0.439718 -0.225971                  0.272104 -0.218759   
3                 -0.609076 -0.230792                  0.396168 -0.220555   
4                  0.216679 -0.155888                 -0.200808 -0.143570   

   30-39 y/o % of total pop  ...  COPD_number  diabetes_prevalence  \
0                  0.588469  ...    -0.138900                 12.9   
1                 -0.110300  ...     0.563986                 12.0   
2                  0.656538  ...    -0.219763                 19.7   
3                  1.264959  ...    -0.256918                 14.1   
4                  0.088582  ...    -0.074198                 13.5   

   diabetes_Lower 95% CI  diabetes_Upper 95% CI  diabetes_number  \
0              -0.071720              -0.089834        -0.129902   
1              -0.414900              -0.337677         0.376251   
2               2.483064               2.317776        -0.183415   
3               0.423984               0.299632        -0.229096   
4               0.195197               0.158008        -0.111247   

   CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  CKD_number  \
0       -0.609615         -0.582796         -0.669652   -0.147523   
1       -0.433549         -0.393279         -0.343373    0.389791   
2        1.855312          1.880929          1.777443   -0.204321   
3       -0.257483         -0.203761         -0.180233   -0.242100   
4       -0.081417         -0.014244         -0.017093   -0.124105   

   Urban_rural_code  
0         -1.082865  
1         -0.420704  
2          0.903618  
3         -1.745026  
4         -1.745026  

[5 rows x 108 columns]

In [150]:
# Divide the dataset into training and test samples
X = df.drop('diabetes_prevalence', axis=1)
y = df['diabetes_prevalence']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 71)

In [154]:
# Feature selection 
train_index = list(X_train.index)
test_index = list(X_test.index)

k = int(len(X_train.columns) * 0.3)
selection_model = SelectKBest(score_func = f_regression, k = k)
selection_model.fit(X_train, y_train)
ix = selection_model.get_support()

X_train_sel = pd.DataFrame(selection_model.transform(X_train), columns = X_train.columns.values[ix])
X_test_sel = pd.DataFrame(selection_model.transform(X_test), columns = X_test.columns.values[ix])

X_train_sel.head()

% White-alone  % Black-alone  R_death_2018  \
0       0.647127      -0.557936     -0.135940   
1       0.757177      -0.565467      1.017714   
2       0.538737      -0.382979     -0.433657   
3      -0.565644      -0.557130     -0.470872   
4      -0.182207       0.125315     -0.024296   

   Percent of adults with less than a high school diploma 2014-18  \
0                                           0.108842                
1                                          -0.506607                
2                                          -0.411923                
3                                          -0.869564                
4                                          -0.711757                

   Percent of adults with a high school diploma only 2014-18  \
0                                           0.154941           
1                                           1.962510           
2                                           1.322909           
3                                          -0.206573           
4                                           0.293985           

   Percent of adults with a bachelor's degree or higher 2014-18  \
0                                          -0.293986              
1                                          -0.421202              
2                                          -0.580221              
3                                           0.289084              
4                                          -0.039556              

   PCTPOVALL_2018  PCTPOV017_2018  PCTPOV517_2018  MEDHHINC_2018  ...  \
0        0.055545        0.246196        0.161446      -0.172093  ...   
1       -0.026009       -0.023228       -0.123384      -0.503825  ...   
2       -0.482712       -0.708012       -0.750012       0.140295  ...   
3       -0.874172       -0.966210       -1.034842       0.653104  ...   
4        0.349140        0.728913        0.605782      -0.313113  ...   

   Heart disease_Lower 95% CI  Heart disease_Upper 95% CI  COPD_prevalence  \
0                    0.800427                    0.857050         0.940367   
1                    0.241092                    0.336239         0.343248   
2                   -0.753281                   -0.705384         0.044689   
3                   -0.442540                   -0.497060        -0.424476   
4                   -0.069650                   -0.184573         0.215294   

   COPD_Lower 95% CI  COPD_Upper 95% CI  diabetes_Lower 95% CI  \
0           0.845366           0.980076               0.195197   
1           0.222683           0.406134              -0.071720   
2           0.031088          -0.014757              -0.338638   
3          -0.352101          -0.435648              -0.643687   
4           0.270582           0.214820              -0.147983   

   diabetes_Upper 95% CI  CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  
0               0.299632        0.622848          0.554308          0.635465  
1               0.051790        0.094650         -0.014244          0.146046  
2              -0.337677       -0.961748         -0.961830         -0.995931  
3              -0.727143       -0.785681         -0.772313         -0.832792  
4              -0.231458       -0.081417         -0.014244         -0.180233  

[5 rows x 32 columns]

In [155]:
X_test_sel.head()

% White-alone  % Black-alone  R_death_2018  \
0      -1.083385       1.414490      0.273421   
1      -0.118298      -0.430712     -0.582516   
2       0.432694      -0.253306     -0.210369   
3       0.635162      -0.540805     -0.917448   
4       0.117249      -0.557066     -1.066306   

   Percent of adults with less than a high school diploma 2014-18  \
0                                           0.661168                
1                                          -1.232521                
2                                           1.892066                
3                                          -0.648634                
4                                          -0.901126                

   Percent of adults with a high school diploma only 2014-18  \
0                                           0.627690           
1                                          -1.624819           
2                                           1.169960           
3                                           0.488646           
4                                          -1.527489           

   Percent of adults with a bachelor's degree or higher 2014-18  \
0                                          -0.622626              
1                                           1.126586              
2                                          -1.555540              
3                                           0.119464              
4                                           1.306808              

   PCTPOVALL_2018  PCTPOV017_2018  PCTPOV517_2018  MEDHHINC_2018  ...  \
0        0.120788        0.156388        0.104480      -0.242208  ...   
1       -0.988348       -1.078470       -1.069022       1.708852  ...   
2        3.024116        2.053578        2.121081      -1.635154  ...   
3       -0.727375       -0.842724       -0.841157       0.681725  ...   
4       -0.042320       -0.752916       -0.772798       0.682804  ...   

   Heart disease_Lower 95% CI  Heart disease_Upper 95% CI  COPD_prevalence  \
0                    0.489685                    0.336239         0.300597   
1                   -1.188319                   -1.226196        -1.320154   
2                    2.602728                    2.419484         3.286191   
3                   -0.628985                   -0.601222        -0.040614   
4                   -1.374764                   -1.330358        -1.277503   

   COPD_Lower 95% CI  COPD_Upper 95% CI  diabetes_Lower 95% CI  \
0           0.366379           0.214820               0.729033   
1          -1.310075          -1.353955              -1.024998   
2           3.431897           3.237582               2.254277   
3          -0.112608          -0.014757              -0.948736   
4          -1.262177          -1.277430              -1.330047   

   diabetes_Upper 95% CI  CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  
0               0.547474        0.798914          0.933342          0.798605  
1              -1.081203       -1.137814         -1.151348         -1.159071  
2               2.176151        1.679246          1.501894          1.614304  
3              -0.904173       -0.785681         -0.772313         -0.832792  
4              -1.399857       -1.137814         -1.151348         -1.159071  

[5 rows x 32 columns]

In [157]:
# Add the target variable "diabetes_prevalence" to the selected features DataFrames
X_train_sel["diabetes_prevalence"] = list(y_train)
X_test_sel["diabetes_prevalence"] = list(y_test)

In [158]:
# Save train and test
X_train_sel.to_csv("/workspaces/ML-Regularized-Linear-Models-pilarzarco/data/processed/clean_train.csv", index = False)
X_test_sel.to_csv("/workspaces/ML-Regularized-Linear-Models-pilarzarco/data/processed/clean_test.csv", index = False)

In [159]:
# Concatenate df X_train and X_test_sel with feature selection
df = pd.concat([X_train_sel, X_test_sel])
df.head()

% White-alone  % Black-alone  R_death_2018  \
0       0.647127      -0.557936     -0.135940   
1       0.757177      -0.565467      1.017714   
2       0.538737      -0.382979     -0.433657   
3      -0.565644      -0.557130     -0.470872   
4      -0.182207       0.125315     -0.024296   

   Percent of adults with less than a high school diploma 2014-18  \
0                                           0.108842                
1                                          -0.506607                
2                                          -0.411923                
3                                          -0.869564                
4                                          -0.711757                

   Percent of adults with a high school diploma only 2014-18  \
0                                           0.154941           
1                                           1.962510           
2                                           1.322909           
3                                          -0.206573           
4                                           0.293985           

   Percent of adults with a bachelor's degree or higher 2014-18  \
0                                          -0.293986              
1                                          -0.421202              
2                                          -0.580221              
3                                           0.289084              
4                                          -0.039556              

   PCTPOVALL_2018  PCTPOV017_2018  PCTPOV517_2018  MEDHHINC_2018  ...  \
0        0.055545        0.246196        0.161446      -0.172093  ...   
1       -0.026009       -0.023228       -0.123384      -0.503825  ...   
2       -0.482712       -0.708012       -0.750012       0.140295  ...   
3       -0.874172       -0.966210       -1.034842       0.653104  ...   
4        0.349140        0.728913        0.605782      -0.313113  ...   

   Heart disease_Upper 95% CI  COPD_prevalence  COPD_Lower 95% CI  \
0                    0.857050         0.940367           0.845366   
1                    0.336239         0.343248           0.222683   
2                   -0.705384         0.044689           0.031088   
3                   -0.497060        -0.424476          -0.352101   
4                   -0.184573         0.215294           0.270582   

   COPD_Upper 95% CI  diabetes_Lower 95% CI  diabetes_Upper 95% CI  \
0           0.980076               0.195197               0.299632   
1           0.406134              -0.071720               0.051790   
2          -0.014757              -0.338638              -0.337677   
3          -0.435648              -0.643687              -0.727143   
4           0.214820              -0.147983              -0.231458   

   CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  diabetes_prevalence  
0        0.622848          0.554308          0.635465                 13.7  
1        0.094650         -0.014244          0.146046                 13.1  
2       -0.961748         -0.961830         -0.995931                 12.2  
3       -0.785681         -0.772313         -0.832792                 11.2  
4       -0.081417         -0.014244         -0.180233                 12.5  

[5 rows x 33 columns]

In [160]:
X_test_sel.head()

% White-alone  % Black-alone  R_death_2018  \
0      -1.083385       1.414490      0.273421   
1      -0.118298      -0.430712     -0.582516   
2       0.432694      -0.253306     -0.210369   
3       0.635162      -0.540805     -0.917448   
4       0.117249      -0.557066     -1.066306   

   Percent of adults with less than a high school diploma 2014-18  \
0                                           0.661168                
1                                          -1.232521                
2                                           1.892066                
3                                          -0.648634                
4                                          -0.901126                

   Percent of adults with a high school diploma only 2014-18  \
0                                           0.627690           
1                                          -1.624819           
2                                           1.169960           
3                                           0.488646           
4                                          -1.527489           

   Percent of adults with a bachelor's degree or higher 2014-18  \
0                                          -0.622626              
1                                           1.126586              
2                                          -1.555540              
3                                           0.119464              
4                                           1.306808              

   PCTPOVALL_2018  PCTPOV017_2018  PCTPOV517_2018  MEDHHINC_2018  ...  \
0        0.120788        0.156388        0.104480      -0.242208  ...   
1       -0.988348       -1.078470       -1.069022       1.708852  ...   
2        3.024116        2.053578        2.121081      -1.635154  ...   
3       -0.727375       -0.842724       -0.841157       0.681725  ...   
4       -0.042320       -0.752916       -0.772798       0.682804  ...   

   Heart disease_Upper 95% CI  COPD_prevalence  COPD_Lower 95% CI  \
0                    0.336239         0.300597           0.366379   
1                   -1.226196        -1.320154          -1.310075   
2                    2.419484         3.286191           3.431897   
3                   -0.601222        -0.040614          -0.112608   
4                   -1.330358        -1.277503          -1.262177   

   COPD_Upper 95% CI  diabetes_Lower 95% CI  diabetes_Upper 95% CI  \
0           0.214820               0.729033               0.547474   
1          -1.353955              -1.024998              -1.081203   
2           3.237582               2.254277               2.176151   
3          -0.014757              -0.948736              -0.904173   
4          -1.277430              -1.330047              -1.399857   

   CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  diabetes_prevalence  
0        0.798914          0.933342          0.798605                 14.8  
1       -1.137814         -1.151348         -1.159071                 10.1  
2        1.679246          1.501894          1.614304                 19.1  
3       -0.785681         -0.772313         -0.832792                 10.6  
4       -1.137814         -1.151348         -1.159071                  9.3  

[5 rows x 33 columns]